src/kernels/raw/finder/interpreterKernelSpecFinderHelper.node.ts

```diff
@@ -34,33 +33,269 @@ import { createPromiseFromCancellation } from '../../../platform/common/cancella
+
+        // Then go through all of the kernels and generate their metadata
+        const distinctKernelMetadata = new Map<string, LocalKernelConnectionMetadata>();
+
+        for (const k of kernelSpecsBelongingToPythonEnvironment) {
+            if (
+                k.language === PYTHON_LANGUAGE &&
+                // Hide default kernel specs only if env variables are empty.
+                // If not empty, then user has modified them.
+                (!k.env || Object.keys(k.env).length === 0) &&
+                isDefaultKernelSpec(k)
+            ) {
+                traceVerbose(
+                    `Hiding default kernel spec '${k.display_name}', '${k.name}', ${getDisplayPath(
+                        k.argv[0]
+                    )} for interpreter ${getDisplayPath(k.interpreterPath)} and spec ${getDisplayPath(k.specFile)}`
+                );
+                continue;
+            }
+            const kernelSpec = isKernelLaunchedViaLocalPythonIPyKernel(k)
+                ? PythonKernelConnectionMetadata.create({
+                      kernelSpec: k,
+                      interpreter: this.interpreter,
+                      id: getKernelId(k, this.interpreter)
+                  })
+                : LocalKernelSpecConnectionMetadata.create({
+                      kernelSpec: k,
+                      interpreter: this.interpreter,
+                      id: getKernelId(k, this.interpreter)
+                  });
+
+            traceVerbose(`Found kernel spec at end of discovery ${kernelSpec?.id}`);
+            // Check if we have already seen this.
+            if (kernelSpec && !distinctKernelMetadata.has(kernelSpec.id)) {
+                distinctKernelMetadata.set(kernelSpec.id, kernelSpec);
+            }
+        }
+
+        // Update spec to have a default spec file
+        const spec = await interpreterSpecificKernelSpec;
+        if (cancelToken.isCancellationRequested) {
+            return [];
+        }
+
+        const result = PythonKernelConnectionMetadata.create({
+            kernelSpec: spec,
+            interpreter: this.interpreter,
+            id: getKernelId(spec, this.interpreter)
+        });
+        traceVerbose(`Kernel for interpreter ${this.interpreter.id} is ${result.id}`);
+        if (!distinctKernelMetadata.has(result.id)) {
+            distinctKernelMetadata.set(result.id, result);
+        }
+        return Array.from(distinctKernelMetadata.values());
+    }
+}
+
+/**
+ * We can have kernel specs in global locations and they could be python.
+ * Some of these could have the python interpreter information in the kernelspec.json file.
+ * Either in the argv or in the metadata (latter being the case where Jupyter extension may have added interpreter info into kernelSpec metadata).
+ */
+export class GlobalPythonKernelSpecFinder implements IDisposable {
```

Comment:
Another new class to clearly separate the concerns
previously when searching through each interpreter ahdn when looking for kernlespecs inside that interrpeter, we would also see whether any of the global kernelspecs matched that interpreter.

That was wrong, 
i.e. if we have 10 interpreters, then 10 times we'd try to match a global kernelspec with each interpeter.
So, if we have 5 global python kernelspecs, then we turn a matche 10*5 = 50 times

Now we only try to match 5 times.